### ISSUE TITLE
_instantiateImageCodec optimization wanted

### ISSUE THREAD

**Issue by taosimple** [flutter/flutter#52802]
## Use case
flutter version: Flutter 1.12.13+hotfix.8
A widget Image width:100 height:100
A picture width:1000 height:1500

1. display the picture into the Image :
`Image.file(new File(path) , fit: BoxFit.cover, cacheWidth: 100, cacheHeight: 100)`
the displayed picture is distorted with size:100*100
2. if I modify as below:
`Image.file(new File(path) , fit: BoxFit.cover, cacheWidth: 100)`
the displayed picture with size:100*150

## Proposal
In 1 the picture is distorted , 2 is waste memory
I wanted is the picture center clip and keep aspect like BoxFit.cover
```
String _instantiateImageCodec(Uint8List list, 
_Callback<Codec> callback, 
_ImageInfo imageInfo, 
int targetWidth, int targetHeight)
  native 'instantiateImageCodec';
```
If codec with a params like Image.fit(BoxFit.cover, BoxFit.fill), the codec image can fit like widget Image, What you see is what you get, is the best.

**Comment by dnfield** [comment]
Yes - see https://flutter.dev/go/encoded-image-dimensions.  The trick is finding some way to make this useful for people with the current `Image` widget API.

**Comment by dnfield** [comment]
I have ideas for this.  Assigning a conservative milestone for now.

**Comment by taosimple** [comment]
@dnfield Thank you for your reply
I have tested `package:image` mentioned in  https://flutter.dev/go/encoded-image-dimensions , It can achieve the purpose by resize and crop, but have 2 problems:

1. Can't work with flutter MemoryCache

2. About 100 times slower than instantiateImageCodec

**Comment by dnfield** [comment]
Yeah, it's a great package but not an really an acceptable solution for this kind of use case.

I think what we need to do is provide some kind of callback or class or image provider that can help out with this - kind of like resize image, but an aspect ratio aware resize image instead.  There are a few possible ways to do it and right now I think it's a matter of figuring out which one would be least breaking and most flexible.

**Comment by dnfield** [comment]
In the mean time, you could probably use `Image.memory` with the code I linked in that document to read the size ahead of time, and do your own work to properly use `cacheHeight` and `cacheWidth` as a work around.

**Comment by dnfield** [comment]
Oh, and I should add for completeness - it would really be best if you just shipped the asset at the right scale.  You're still going to use more CPU and memory (and network or disk bandwidth) to scale the image on device than if it was just decoded at the right size and scale to begin with.

**Comment by taosimple** [comment]
> Oh, and I should add for completeness - it would really be best if you just shipped the asset at the right scale. You're still going to use more CPU and memory (and network or disk bandwidth) to scale the image on device than if it was just decoded at the right size and scale to begin with.

My English is not good. I'm not sure what you mean by messages above.I'm just use local image files, so don't need to use the network, the main impact is memory usage and performance

**Comment by dnfield** [comment]
Can you just resize the images ahead of time, and include images that are the correct size/scale in your application bundle?

**Comment by taosimple** [comment]
> Can you just resize the images ahead of time, and include images that are the correct size/scale in your application bundle?

In principle, it is feasible, but the processing is more complicated, which needs create a map between thumbnail and src image.

**Comment by taosimple** [comment]
> Can you just resize the images ahead of time, and include images that are the correct size/scale in your application bundle?

I referenced the principle of the [Glide](https://github.com/bumptech/glide), I think i should resize the images ahead of time and generate thumbnail, Thanks again.

**Comment by dnfield** [comment]
Changes needed in the engine have landed for this. I'm going to look at how we can make ResizeImage better utilize them.

**Comment by MatrixDev** [comment]
I will leave it here in case someone needs this in the future:

```dart
import 'dart:typed_data';
import 'dart:ui' as ui show Codec, ImmutableBuffer, ImageDescriptor;

import 'package:flutter/foundation.dart';
import 'package:flutter/painting.dart';

class AspectResizeImageProvider extends ImageProvider<String> {
  final int width;
  final int height;
  final ImageProvider provider;

  AspectResizeImageProvider({
    @required this.width,
    @required this.height,
    @required this.provider,
  });

  @override
  ImageStreamCompleter load(String key, DecoderCallback decode) {
    return provider.load(key, _smartDecode);
  }

  Future<ui.Codec> _smartDecode(
    Uint8List bytes, {
    int cacheWidth,
    int cacheHeight,
    bool allowUpscaling,
  }) async {
    final buffer = await ui.ImmutableBuffer.fromUint8List(bytes);
    final descriptor = await ui.ImageDescriptor.encoded(buffer);

    int sourceWidth = descriptor.width;
    int sourceHeight = descriptor.height;
    if (sourceWidth == null || sourceHeight == null) {
      // sometimes ImageDescriptor doesn't return width/height
      // will manually decode image to get real image dimensions
      final codec = await descriptor.instantiateCodec();
      try {
        final frame = await codec.getNextFrame();
        sourceWidth = frame.image.width;
        sourceHeight = frame.image.height;
      } finally {
        codec.dispose();
      }
    }

    int targetWidth;
    int targetHeight;
    if (sourceWidth != null && sourceHeight != null) {
      // center crop with upscaling is used here. change implementation if needed
      final targetAspect = width / height;
      final sourceAspect = sourceWidth / sourceHeight;
      if (targetAspect > sourceAspect) {
        targetHeight = height;
        targetWidth = (height * sourceAspect).round();
      } else {
        targetWidth = width;
        targetHeight = (width / sourceAspect).round();
      }
    }

    return descriptor.instantiateCodec(
      targetWidth: targetWidth,
      targetHeight: targetHeight,
    );
  }

  // TODO obtainKey implementation, not relevant to my comment
}
```
